### PR TITLE
Is pattern expression with in parameter avoid unnecesary temp in rewriter

### PR DESCRIPTION
**Pull request by ArcadeMode**
Fix for #65091

### CODE REVIEW

**Comment by ArcadeMode** [comment]
The failing checks appear related to some runs timing out after 90 minutes, not sure if thats related to my changes

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenSystemNullableTests.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        [WorkItem(65091, "https://github.com/dotnet/roslyn/issues/65091")]
+        public void Nullable_Parameter_EqualsNull()
+        {
+            var source = @"
+void M(in int? v)
+{
+    if (v == null) 
+        return;
+}";
+            var verifier = CompileAndVerify(source);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenSystemNullableTests.cs:28]: >CompileAndVerify(source) [](http://example.com/codeflow?start=27&length=24)

Please update the new passing tests to also check execution (`CompileAndVerify(..., expectedOutput: ...)`) using `Console.Write(...)` in the source of the compilation.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenSystemNullableTests.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
+{
+    public class CodeGenSystemNullableTests : CSharpTestBase
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenSystemNullableTests.cs:16]: >CodeGenSystemNullableTests [](http://example.com/codeflow?start=17&length=26)

It doesn't feel like we need a new test file just for this. Can you add the tests into `src\Compilers\CSharp\Test\Emit2\Semantics\PatternMatchingTests5.cs` instead?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenSystemNullableTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenSystemNullableTests.cs:28]: Also, since the three tests are equivalent (same IL), consider consolidating into one with a `[Theory]`, using something like `[InlineData("v == null")][InlineData("!v.HasValue")]...`

**Review by jcouv — COMMENTED** [review]
Change looks good, only minor feedback on the tests :-) 
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs [hunk 3]
```diff
@@ -516,8 +516,8 @@ private BoundExpression MakeSpanStringTest(BoundExpression input, ConstantValue
                     .Any(static node => node is BoundWhenDecisionDagNode { WhenExpression: { ConstantValue: null } });
 
                 var inputDagTemp = BoundDagTemp.ForOriginalInput(loweredInput);
-                if ((loweredInput.Kind == BoundKind.Local || loweredInput.Kind == BoundKind.Parameter)
-                    && loweredInput.GetRefKind() == RefKind.None &&
+                if (loweredInput.Kind is BoundKind.Local or BoundKind.Parameter &&
+                    loweredInput.GetRefKind() is RefKind.None or RefKind.In &&
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:520]: we should test a scenario where a "mutating" member is accessed in the pattern

```cs
void M(in S s)
{
    if (s.Value is 1) { Console.Write(1); }
    if (s.Value is 1) { Console.Write(2); }
}

struct S
{
    int item;
    public int Value => ++item;
}
```

I am pretty sure I would expect implicit copies to be made here and to get output of `12`.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:520]: See the comment below (line 523). The language and compiler assume that functions called by pattern-matching are not side-effecting.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:520]: Discussed offline. We agree that when the argument is a readonly reference (i.e. `in`), and the parameter on the method/accessor is a writable reference (i.e. `ref`), then we still need to make sure we copy the value and pass a writable reference to the copy, in order to preserve readonly safety.

I think we should try changing the condition to try and identify/filter out that specific scenario. Unfortunately, since the access could happen anywhere in the DAG, we might have to go ahead and do a DAG traversal (maybe a partial traversal?) to decide if a writable-reference member is ever accessed on the original input.

First step is definitely to just add the test, and particularly the flavor with `s is { Value: 1 }` repeated, to determine whether the current implementation results in a readonly safety violation.

Let me know if this makes sense @ArcadeMode, I'm happy to assist further if I can.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:520]: > or RefKind.In [](http://example.com/codeflow?start=61&length=14)

I am not sure why does this matter. The `in` modifier on the parameter doesn't provide a guarantee that the value doesn't change. It prevents this particular function to make the change.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs:520]: If, for example, we could detect that there is the only check for this local in the entire Dag, then we could avoid doing copy even for `ref` locals/parameters.
Alternatively, we could say that we do not care about changes made by other threads, and, given that we already assume that pattern matching is not side-effecting, we could careless about ref-ness and provide "good" code gen even for `ref` case.
Bottom line, I am not following why 'in' is somehow special here.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 4)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs (hunk 3 above)
Comment (ArcadeMode) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs]: I have added the mentioned test cases and  readonly safety appears to be okay.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs]: Could you give a concrete example where doing the same optimization for `ref` is going to give an incorrect result?

Comment (ArcadeMode) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs]: So if we take the following example, readonly safety is preserved in the current implementation, ie the output is "12"
```csharp
M(default(S));

void M(in S s)
{
    if (s.Value is 1) { Console.Write(1); }
    if (s.Value is 1) { Console.Write(2); }
}

struct S
{
    int item;
    public int Value => ++item;
}
```

If we do the same with a ref parameter, readonly safety is violated and the output is "1"
```csharp
var s = default(S);
M(ref s);

void M(ref S s)
{
    if (s.Value is 1) { Console.Write(1); }
    if (s.Value is 1) { Console.Write(2); }
}

struct S
{
    int item;
    public int Value => ++item;
}
```
(I have tested the scenarios locally)


Now these examples technically do violate the assumption of idempotency (no side-effecting), so we could optimize the `ref` case anyway, as you commented above. I will leave this decision up to the roslyn team.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs]: I am not sure if this is good example. We already observe this difference in behavior today, even without an optimization performed in this PR. And as far as I can see, the 'ref' parameter is not getting copied today.
```
    static void M_Ref(ref S s)
    {
        if (s.Value is 1) { Console.WriteLine(1); }
    }
```

The corresponding IL is:
```
    .method private hidebysig static 
        void M_Ref (
            valuetype S& s
        ) cil managed 
    {
        .maxstack 2
        .locals init (
            [0] bool
        )

        IL_0000: nop
        IL_0001: ldarg.0
        IL_0002: call instance int32 S::get_Value()
        IL_0007: ldc.i4.1
        IL_0008: ceq
        IL_000a: stloc.0
        IL_000b: ldloc.0
        IL_000c: brfalse.s IL_0017
        IL_000e: nop
        IL_000f: ldc.i4.1
        IL_0010: call void [System.Console]System.Console::WriteLine(int32)
        IL_0015: nop
        IL_0016: nop
        IL_0017: ret
    } // end of method Program::M_Ref
```

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs]: I do not expect optimization to have any effect on this scenario.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs]: @ArcadeMode

> I will leave this decision up to the roslyn team.

 We discussed this offline and we fill comfortable with removing the ref kind check here completely. Let's go with that for now.

Unfortunately it looks like this change is not sufficient for improving code generated for this scenario:
```
class C
{
    static void PoorCodegen1(S s)
    {
        if (s.F is null) 
            System.Console.WriteLine();
    }
}

struct S
{
    public int? F;
}
```

It feels to me that we would want to improve this and the like scenarios as well. Could you investigate?

**Comment by AlekseyTs** [comment]
@ArcadeMode What is the status of this PR?

**Comment by RikkiGibson** [comment]
Marking as draft for now. We'd be interested in looking more closely at how exactly this fix should work. Feel free to follow up with us here or in the Roslyn channel in https://discord.gg/csharp.
